### ISSUE TITLE
Merge Duplicates

### ISSUE THREAD

**Issue by troughc** [culturecreates/footlight-app#1200]
First: sorry if this ticket already exists - I could not find it in our board.

Final LINK
https://www.figma.com/design/DBnSSMT9Etfm6guDKKSp8O/Culture-Creates---Footlight-Calendar?node-id=10157-100372&m=dev

New feature for Admins only to merge duplicates. This feature will be available in all workspaces lists (events, people, places, organizations). The Figma illustrates the events view as an example/model for how to implement
Of note: the checkboxes (left of the thumbnail image in the list views) will always be displayed for Admins. The checkboxes are not displayed for other user profiles.

French translations have been added in FIGMA

Tasklist : 

Events:

- [ ] 1. Make the event listing page selectable.
- [ ] 2. Introduce Merge duplicates and cancel button in the footer.Create the footer component to reuse. 
- [ ] 3. Add routing to event forms from the listing pages.
- [ ] 4. Api integration to get the details of selected events.
- [ ] 5. Create the number component to show the no.of duplicates found and add dropdown.
- [ ] 6. Add "review merged content" component to the right.
- [ ] 7. Add the logic to show "review merged content " or "remove unwanted versions".
- [ ] 8. Compare the values in the respective fields show each and every field where there is a duplicate.(The selected field can also be edited).
- [ ] 9. If image duplicates exists ,they should be compared based on the Uri.
- [ ] 10. Add the date component to show duplicates.
- [ ] 11. Add the tree selected component to show duplicates.
- [ ] 12. Add the api integration to update the respective event.
- [ ] 13. Add the api integrations to delete unwanted events.

People :

- [ ] 1. Make the people listing page selectable.
- [ ] 2. Reuse Merge duplicates and cancel button in the footer. 
- [ ] 3. Add routing to person forms from the listing pages.
- [ ] 4. Api integration to get the details of selected person.
- [ ] 5. Integration to show the duplicates found.
- [ ] 6. Add "review merged content" component to the right.
- [ ] 7. Add the logic to show "review merged content " or "remove unwanted versions".
- [ ] 8. Compare the values in the respective fields show each and every field where there is a duplicate.(The selected field can also be edited).
- [ ] 9. If image duplicates exists ,they should be compared based on the Uri.
- [ ] 10. Add the tree selected component to show duplicates.
- [ ] 11. Add the api integration to update the respective person.
- [ ] 12. Add the api integrations to delete unwanted people.
- [ ] 13. Show the social media link.Let the user choose to remove the link.

Place :

- [ ] 1. Make the Place listing page selectable.
- [ ] 2. Reuse Merge duplicates and cancel button in the footer. 
- [ ] 3. Add routing to place form.
- [ ] 4. Api integration to get the details of selected place.
- [ ] 5. Integration to show duplicate found.
- [ ] 6. Add "review merged content" component to the right.
- [ ] 7. Add the logic to show "review merged content " or "remove unwanted versions".
- [ ] 8.  Compare the values in the respective fields show each and every field where there is a duplicate.(The selected field can also be edited).
- [ ] 9.  If image duplicates exists ,they should be compared based on the Uri.
- [ ] 10.  Add the tree selected component to show duplicates.
- [ ] 11. Compare the street name,city,province ,country,coordinates and update the respective postal address.
- [ ] 12. Add the api integration to update the respective place
- [ ] 13. Add the api integrations to delete unwanted people.

Organization : 

- [ ] 1. Make the organization listing page selectable.
- [ ] 2. Reuse Merge duplicates and cancel button in the footer. 
- [ ] 3. Add routing to organization form.
- [ ] 4. Api integration to get the details of selected organization.
- [ ] 5. Integration to show duplicate found.
- [ ] 6. Add "review merged content" component to the right.
- [ ] 7. Add the logic to show "review merged content " or "remove unwanted versions".
- [ ] 8.  Compare the values in the respective fields show each and every field where there is a duplicate.(The selected field can also be edited).
- [ ] 9.  If image and logo duplicates exists ,they should be compared based on the Uri.
- [ ] 10. Add the tree selected component to show duplicates.
- [ ] 11. Add the api integration to update the respective organization
- [ ] 12. Add the api integrations to delete unwanted organization.

**Comment by troughc** [comment]
@saumier @sahalali here is the issue for the merge duplicates feature

**Comment by sahalali** [comment]
@troughc 
Please take note of the following points:
1. How should we compare events to find duplicates? For instance, should duplicates be identified based on Artsdata ids, titles, dates, or a combination of these?
2. When merging two or more events, the priority for which event should be preserved. For example, if event A and event B are merged, the priority associated with event A should be retained, and event B should be deleted.
3. It seems necessary to enable the selection of events from multiple pages as users may find it difficult to filter events for merging within the same page.
4. Users should have the option to select events based on multiple searches. For instance, they should be able to merge events with entirely different search parameters.
5. An option to deselect events should be included.
6. Should users be restricted from merging events in the Published state, or merging Draft events with Published events?
7. Should the merge option be restricted based on user roles?
8. I think it will be helpful for the user if a 'tag' can be added to each value to differentiate the source event. It can be a message on hover or a coloured tag.

**Comment by AbhishekPAnil** [comment]
Frontend estimates : 

Events:

- [ ] 1. Make the event listing page selectable.
- [ ] 2. Introduce Merge duplicates and cancel button in the footer.Create the footer component to reuse. 
- [ ] 3. Add routing to event forms from the listing pages.
- [ ] 4. Api integration to get the details of selected events.
- [ ] 5. Create the number component to show the no.of duplicates found and add dropdown.
- [ ] 6. Add "review merged content" component to the right.
- [ ] 7. Add the logic to show "review merged content " or "remove unwanted versions".
- [ ] 8. Compare the values in the respective fields show each and every field where there is a duplicate.(The selected field can also be edited).
- [ ] 9. If image duplicates exists ,they should be compared based on the Uri.
- [ ] 10. Add the date component to show duplicates.
- [ ] 11. Add the tree selected component to show duplicates.
- [ ] 12. Add the api integration to update the respective event.
- [ ] 13. Add the api integrations to delete unwanted events.

**Comment by troughc** [comment]
@AbhishekPAnil I would like to get feedback from @saumier when he returns, but I am assigning this back to you. thanks!

**Comment by saumier** [comment]
> If event A has an Artsdata ID and Event B does not, I think that the merged date title and location would have to match the Artsdata ID event for it to be the event which is retained. If it does not match, I think you should save the merged event as a a new event (without an Artsdata ID). This new event would get a new Artsdata ID. @saumier is this correct?
> If events A, B , C... all have DIFFERENT Artsdata IDs, only one would be preserved in the CMS if the user chooses to merge (and save) them all (the user may choose NOT to merge). As for which Artsdata ID to choose: I think the one that most closely matches (same date, title, location, and maybe other properties like event type, description, performer/organizer). @saumier can you please help with this?

In all cases of events resulting from a merge, I think the Artsdata ID or IDs in CMS should be wiped out (removed). There has been an error somewhere in Artsdata for the duplicates to happen in the first place, and since there is no visibility in CMS of the ID of Artsdata IDs I don't think CMS should be responsible for choosing the right Artsdata ID.  Let Artsdata relink or mint on its own.

**Comment by AbhishekPAnil** [comment]
cc @AbhishekPAnil For this pass add select all checkbox to the top of the listing pages.

**Comment by AbhishekPAnil** [comment]
@troughc Should a user be able to share the list of selected duplicates ?Like the event listing with the selected filters we have right now!

**Comment by troughc** [comment]
One thing to mention @AbhishekPAnil @dev-aravind: remember that when an Admin merges duplicates that are places organizations or people, the final entity needs to link to ALL the other entities that each duplicate was linked to.
For example if you merge two places because they are duplicates, all events that have either place as a location should continue to be linked to the final, chosen entity.
Let me know if its not clear, we can speak about it next week.

**Comment by AbhishekPAnil** [comment]
@troughc How do you want the select behaviour to be, when a user applies the filters in the listing pages?What happens to selected entities when a user clears the filters ? Should it persist in the respective page numbers !

**Comment by troughc** [comment]
As per our discussion on the above questions: I am fine with the approach you are taking. If we need to revisit later, we can.

**Comment by AbhishekPAnil** [comment]
Conclusions from the discussion:

>How do you want the select behaviour to be, when a user applies the filters in the listing pages?What happens to selected entities when a user clears the filters ? Should it persist in the respective page numbers?

The selected events should persist even if the filters are cleared.This might be hard to achieve but cc @AbhishekPAnil will find a solution for this.

>Can we merge past events while doing merging duplicates ?

Yes

**Comment by AbhishekPAnil** [comment]
> One thing to mention [@AbhishekPAnil](https://github.com/AbhishekPAnil) [@dev-aravind](https://github.com/dev-aravind): remember that when an Admin merges duplicates that are places organizations or people, the final entity needs to link to ALL the other entities that each duplicate was linked to. For example if you merge two places because they are duplicates, all events that have either place as a location should continue to be linked to the final, chosen entity.
> 
> Question: is this what you are doing?
> 
> cc [@sahalali](https://github.com/sahalali) [@saumier](https://github.com/saumier)

This has been noted .

**Comment by dev-aravind** [comment]
Task for @dev-aravind:

Modify all the entities update APIs to remap deleted entity references to the merged one.

**Comment by dev-aravind** [comment]
@AbhishekPAnil The backend changes can be found in [this](https://github.com/culturecreates/footlight-calendar-api/tree/feature/issue-1200) branch. Please test it and let me know if you need any changes.

**Comment by troughc** [comment]
@AbhishekPAnil I replied to your question in the Figma.

**Comment by AbhishekPAnil** [comment]
cc @AbhishekPAnil Will be displaying the dates separately even if the events have same dates.

**Comment by sahalali** [comment]
@AbhishekPAnil What is the status of this ticket?

**Comment by AbhishekPAnil** [comment]
> @AbhishekPAnil What is the status of this ticket?
> 

This is in progress. Currently working on integrating the language changes that syam did.I am doing this ticket in parallel.

**Comment by sahalali** [comment]
> > If event A has an Artsdata ID and Event B does not, I think that the merged date title and location would have to match the Artsdata ID event for it to be the event which is retained. If it does not match, I think you should save the merged event as a a new event (without an Artsdata ID). This new event would get a new Artsdata ID. [@saumier](https://github.com/saumier) is this correct?
> > If events A, B , C... all have DIFFERENT Artsdata IDs, only one would be preserved in the CMS if the user chooses to merge (and save) them all (the user may choose NOT to merge). As for which Artsdata ID to choose: I think the one that most closely matches (same date, title, location, and maybe other properties like event type, description, performer/organizer). [@saumier](https://github.com/saumier) can you please help with this?
> 
> [@troughc](https://github.com/troughc) [@sahalali](https://github.com/sahalali) In all cases of entities resulting from a merge, I think the Artsdata IDs in CMS that are different should be added to the merged entity without any special consideration. It will help traceability. This is also true for sameAs URIs that are not visible to the user in the UI and cannot be selected by the user (like Footlight Console URIs, Caligram URIs, ...). Even if an event has 2 or 3 or 4 ... Artsdata IDs after the merge. For there to be more than 1 Artsdata ID there has been an error somewhere in Artsdata. I don't think CMS should be responsible for choosing the right Artsdata ID and there is no visibility in CMS of the ID of Artsdata IDs which makes it currently impossible for a user to select one. Let Artsdata be responsible to link or mint on its own and create the sameAs between Artsdata IDs if needed. In the CMS UI that has a link to Artsdata, it can link to the first arbitray Artsdata ID if there is more than one Artsdata ID in the merged event.

@saumier I have question, what if the event date / title is mismatching with the Artsdata entity. 
Example: If the merged entity doesn't match the date or title in Artsdata after merge.

**Comment by AbhishekPAnil** [comment]
cc @dev-aravind 
Please add support for

1. making rest of the merged events obsolete.
2. merging events into one event.

**Comment by saumier** [comment]
> @saumier I have question, what if the event date / title is mismatching with the Artsdata entity.
> Example: If the merged entity doesn't match the date or title in Artsdata after merge.

@sahalali My opinion has not changed. I still think the merge should not worry about anything else but merging. Since the CMS has no UI to edit or change sameAs URIs there should not be any business logic in how to treat them other than merge.  When the data is loaded into Artsdata it will be handled there.  In most cases the script that links and unlinks CMS to Artsdata will take care of it. That script is where the business logic should be focused on and only there.  The other advantage of merging all sameAs identifiers that that it will appear in the logs and change history even if it gets unlinked in the next step.

**Comment by AbhishekPAnil** [comment]
Cc @AbhishekPAnil Add the remaining estimates for the event -merge duplicates and create separate tickets for the other workspaces and re-estimate the same.

**Comment by AbhishekPAnil** [comment]
Remaining estimates :  24-28 hrs.

**Comment by AbhishekPAnil** [comment]
cc @AbhishekPAnil Hold off on this issue and work on #1449

**Comment by sahalali** [comment]
@AbhishekPAnil  
Could you provide us with an update on the status of this issue? We understand that our focus is on completing the event screen in this phase. Additionally, can you help us estimate how many more hours of effort will be needed to complete phase 1?

**Comment by AbhishekPAnil** [comment]
Remaining tasks : 23h
1.Add the logic to merge the fields - 6h
2.Show warnings if the there are unmerged fields - 3h
3.Align warning the the fields css - 3h
4.Payload integration for the merged events - 4h
4.Testing - 7h

**Comment by sahalali** [comment]
We can add this to the sprint in July.

**Comment by AbhishekPAnil** [comment]
Key Takeaways from the Discussion – 05 August 2025

1. Remove the warning messages currently displayed on the right side for all fields.
2. Eliminate the existing conflict logic specific to the taxonomy multi-select field and align its behavior with how conflicts are handled for other fields.
3. Display a unified warning message for all conflicts at the top of the merge page using a yellow banner
4. Show the total count of conflicting fields alongside the banner, and update the counter as users resolve each conflict.
5. Conflict fields should remain disabled until they are resolved; once resolved, users should be allowed to edit those fields.
6. After all conflicts are resolved, and the user chooses to merge and either publish or save as draft, provide an option to merge into an existing selected event or create a new one.
 Note: @AbhishekPAnil and @sahalali will discuss how to handle scenarios involving a combination of published and draft events.
7. @AbhishekPAnil to revisit and re-estimate the effort for this issue.

**Comment by AbhishekPAnil** [comment]
@dev-aravind
Please update the existing backend branch for merge duplicates with the latest changes, including the Node updates. Also, review the current implementation for point number 6 mentioned above and confirm the approach with @sahalali.

**Comment by sahalali** [comment]
It would be helpful to enable merged duplicate notification to event owners to eliminate confusion.

**Comment by troughc** [comment]
FYI :Due to a higher demand than expected for the work in the Artsdata sprint, this ticket was postponed to next sprint.

**Comment by AbhishekPAnil** [comment]
> [@sahalali](https://github.com/sahalali) [@dev-aravind](https://github.com/dev-aravind) Please update the delete event api to accept multiple ids .

Please update this and assign the ticket back to me. @dev-aravind
